scipy/spatial/distance.pyi

```diff
@@ -0,0 +1,222 @@
+import sys
+from typing import overload, Optional, Any, Union, Tuple, SupportsFloat
+
+import numpy as np
+from numpy.typing import ArrayLike
+
+if sys.version_info >= (3, 8):
+    from typing import Literal, Protocol, SupportsIndex
+else:
+    from typing_extensions import Literal, Protocol
+
+# Anything that can be parsed by `np.float64.__init__` and is thus
+# compatible with `ndarray.__setitem__` (for a float64 array)
+if sys.version_info >= (3, 8):
+    _FloatValue = Union[None, str, bytes, SupportsFloat, SupportsIndex]
+else:
+    _FloatValue = Union[None, str, bytes, SupportsFloat]
+
+class _MetricCallback1(Protocol):
+    def __call__(
+        self, __XA: np.ndarray, __XB: np.ndarray
+    ) -> _FloatValue: ...
+
+class _MetricCallback2(Protocol):
+    def __call__(
+        self, __XA: np.ndarray, __XB: np.ndarray, **kwargs: Any
+    ) -> _FloatValue: ...
+
+# TODO: Use a single protocol with a parameter specification variable
+# once available (PEP 612)
+_MetricCallback = Union[_MetricCallback1, _MetricCallback2]
+
+_MetricKind = Literal[
+    'braycurtis',
+    'canberra',
+    'chebychev', 'chebyshev', 'cheby', 'cheb', 'ch',
+    'cityblock', 'cblock', 'cb', 'c',
+    'correlation', 'co',
+    'cosine', 'cos',
+    'dice',
+    'euclidean', 'euclid', 'eu', 'e',
+    'matching', 'hamming', 'hamm', 'ha', 'h',
+    'minkowski', 'mi', 'm', 'pnorm',
+    'jaccard', 'jacc', 'ja', 'j',
+    'jensenshannon', 'js',
+    'kulsinski',
+    'mahalanobis', 'mahal', 'mah',
+    'rogerstanimoto',
+    'russellrao',
+    'seuclidean', 'se', 's',
+    'sokalmichener',
+    'sokalsneath',
+    'sqeuclidean', 'sqe', 'sqeuclid',
+    # NOTE: deprecated
+    # 'wminkowski', 'wmi', 'wm', 'wpnorm',
+    'yule',
+]
+
+# Function annotations
+
+def braycurtis(
+    u: ArrayLike, v: ArrayLike, w: Optional[ArrayLike] = ...
+) -> np.float64: ...
+
+def canberra(
+    u: ArrayLike, v: ArrayLike, w: Optional[ArrayLike] = ...
+) -> np.float64: ...
+
+# TODO: Add `metric`-specific overloads
+@overload
+def cdist(
+    XA: ArrayLike,
+    XB: ArrayLike,
+    metric: _MetricKind = ...,
+    *,
+    out: Optional[np.ndarray] = ...,
+    p: float = ...,
+    w: Optional[ArrayLike] = ...,
+    V: Optional[ArrayLike] = ...,
+    VI: Optional[ArrayLike] = ...,
+) -> np.ndarray: ...
+@overload
+def cdist(
+    XA: ArrayLike,
+    XB: ArrayLike,
+    metric: _MetricCallback,
+    *,
+    out: Optional[np.ndarray] = ...,
+    **kwargs: Any,
+) -> np.ndarray: ...
+
+def chebyshev(
+    u: ArrayLike, v: ArrayLike, w: Optional[ArrayLike] = ...
+) -> np.float64: ...
```

Comment:
Oh hang on I've found something: both `chebyshev` and `cityblock` do not forcefully cast their inputs into `float64` arrays. As a consequence, they can return any scalar ranging from `uint8` to `float128` (and anything in between). https://github.com/scipy/scipy/blob/d02a18605714180550bc4287f3964699d6466517/scipy/spatial/distance.py#L1040-L1046

For now I'd propose to change the return type to `Any`, as this means we won't have to mess with returning unsafe unions. That, and perhaps update the docstring, as both functions currently claim to return a double.

Any thoughts?

Reply:
That sounds good to me.

Reply:
Done as of https://github.com/scipy/scipy/pull/13448/commits/c2ac950c853cdadd2e0f84b7346eff6ca886d823.